benchmarks/manoptuna.jl

```diff
@@ -0,0 +1,215 @@
+    retr = objective.retrs[pyconvert(
+        Int,
+        trial.suggest_categorical("retraction_method", Vector(eachindex(objective.retrs))),
+    )]
+
+    manifold_name, manifold_constructor = objective.manifold_constructors[pyconvert(
+        Int,
+        trial.suggest_categorical(
+            "manifold", Vector(eachindex(objective.manifold_constructors))
+        ),
+    )]
+
+    manopt_stepsize_name, manopt_stepsize_constructor = objective.manopt_stepsize[pyconvert(
+        Int,
+        trial.suggest_categorical(
+            "manopt_stepsize", Vector(eachindex(objective.manopt_stepsize))
+        ),
+    )]
+
+    local c1_val, c2_val, hz_sigma
+    if manopt_stepsize_name == "Wolfe-Powell"
+        c1_val = pyconvert(
+            Float64, trial.suggest_float("Wolfe-Powell c1", 1e-5, 1e-2; log=true)
+        )
+        c2_val =
+            1.0 - pyconvert(
+                Float64, trial.suggest_float("Wolfe-Powell 1-c2", 1e-4, 1e-2; log=true)
+            )
+    elseif manopt_stepsize_name == "Improved HZ"
+        hz_sigma = pyconvert(Float64, trial.suggest_float("Improved HZ sigma", 0.1, 0.9))
+    end
+
+    loss = sum(objective.pruning_losses)
+
+    # here iterate over problems we want to optimize for
+    # from smallest to largest; pruning should stop the iteration early
+    # if the hyperparameter set is not promising
+    cur_i = 0
+    for N in objective.N_range
+        x0 = zeros(N)
+        x0[1] = 1
+        M = manifold_constructor(N)
+        local ls
+        if manopt_stepsize_name == "Wolfe-Powell"
+            ls = manopt_stepsize_constructor(M, c1_val, c2_val)
+        elseif manopt_stepsize_name == "Improved HZ"
+            ls = manopt_stepsize_constructor(M, hz_sigma)
+        else
+            ls = manopt_stepsize_constructor(M)
+        end
+        manopt_time, manopt_iters, manopt_obj = benchmark_time_state(
+            ManoptQN(),
+            M,
+            N,
+            objective.obj,
+            objective.grad,
+            x0,
+            ls,
+            pyconvert(Int, mem_len),
+            objective.gtol;
+            vector_transport_method=vt,
+            retraction_method=retr,
+        )
+        # TODO: turn this into multi-criteria optimization when Optuna starts supporting
+        # pruning in such problems
+        loss -= objective.pruning_losses[cur_i + 1]
+        loss += manopt_time + objective.obj_loss_coeff * manopt_obj
+        trial.report(loss, cur_i)
+        if pyconvert(Bool, trial.should_prune().__bool__())
+            throw(PyException(optuna.TrialPruned()))
+        end
+        cur_i += 1
+    end
+    return loss
+end
+
+"""
+    lbfgs_study(; pruning_coeff::Float64=0.95)
+
+ensure `pruning_losses` is actually somewhat realistic,
+otherwise there is too little pruning (if values here are too high)
+or too much pruning (if values here are too low)
+regenerate using `lbfgs_compute_pruning_losses`
+"""
+function lbfgs_study(; pruning_coeff::Float64=0.95)
```

Comment:
This also looks quite technical, so.I am not so sure what it does and what its purpose in the end is?
Should this later be a tutorial?